### PR TITLE
PFW-1212: Add Dutch language

### PR DESCRIPTION
**Pull request by 3d-gussner**
This is an attempt to add additional languages to Prusa Firmware, starting with Dutch.

Why: As we have in our maker space in Belgium children and few people not that familiar with English and there are other Dutch speaking people might appreciate this, I started to translate the Prusa Firmware in June 2017 for the Prusa MK2. After some feedback from users we added French and Portuguese as well. The Pull Requests in the past didn't make it BUT i hope Prusa might consider to implement more languages. 

I also wrote basic `How-to add a new language to Prusa Firmware` so others may be able to add other languages.

PFW-1212

### CODE REVIEW

File: lang/po/Firmware_nl.po [hunk 1]
```diff
@@ -0,0 +1,1787 @@
...
+"Language-Team: \n"
+"X-Generator: Poedit 2.0.7\n"
+"X-Poedit-SourceCharset: UTF-8\n"
+"Last-Translator: \n"
+"Plural-Forms: nplurals=3; plural=(n==1) ? 0 : (n>=2 && n<=4) ? 1 : 2;\n"
+
+# 
+#: 
+msgid "[%.7s]Live adj. Z\x0avalue set, continue\x0aor start from zero?\x0a%cContinue%cReset"
+msgstr "[%.7s]Z instell.\x0awaarde ingesteld,\x0averder of bij 0\x0abeginen?\x0a%cVerder%cReset"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po:21]: "terugvallen op 0" might be better than "bij 0 beginnen" for the intent of "revert to 0"

File: lang/po/Firmware_nl.po [hunk 2]
```diff
@@ -0,0 +1,1787 @@
...
+
+# MSG_AUTOLOAD_FILAMENT c=17
+#: ultralcd.cpp:6928
+msgid "AutoLoad filament"
+msgstr "Auto-Laden Filament"
+
+# MSG_AUTOLOADING_ONLY_IF_FSENS_ON c=20 r=4
+#: ultralcd.cpp:4459
+msgid "Autoloading filament available only when filament sensor is turned on..."
+msgstr "Autoloading filament alleen beschikbaar wanneer filamentsensor is ingeschakeld ..."
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po]: "Auto-laden" for consistency with the previous?

File: lang/po/Firmware_nl.po [hunk 3]
```diff
@@ -0,0 +1,1787 @@
...
+
+# MSG_BED_HEATING
+#: messages.c:17
+msgid "Bed Heating"
+msgstr "Bed opwarmen"
+
+# MSG_BED_CORRECTION_MENU
+#: ultralcd.cpp:5838
+msgid "Bed level correct"
+msgstr "Bed niveau correct"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po:166]: "vlak" of "waterpas" is better for "flatness" in the context of bed levelling; "niveau" is generally more used in the context of height.

File: lang/po/Firmware_nl.po [hunk 4]
```diff
@@ -0,0 +1,1787 @@
...
+
+# MSG_EJECT_FILAMENT c=17 r=1
+#: messages.c:53
+msgid "Eject filament"
+msgstr "Fil. uitwerpen"
+
+# MSG_EJECTING_FILAMENT c=20 r=1
+#: mmu.cpp:1413
+msgid "Ejecting filament"
+msgstr "Werp filament uit"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po]: Translation seems closer to "remove filament" than ejecting, if there's space I'd suggest
"Fil. aan het uitwerpen" ("filament is ejecting")

File: lang/po/Firmware_nl.po [hunk 5]
```diff
@@ -0,0 +1,1787 @@
...
+
+# MSG_WIZARD_V2_CAL c=20 r=8
+#: ultralcd.cpp:4957
+msgid "Now I will calibrate distance between tip of the nozzle and heatbed surface."
+msgstr "Begin met kalibratie tussen de tuit en het bed."
+
+# MSG_WIZARD_WILL_PREHEAT c=20 r=4
+#: ultralcd.cpp:5080
+msgid "Now I will preheat nozzle for PLA."
+msgstr "Opwarmen van de tuit voor PLA voor."
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po:921]: "Voorverwarming" is probably better for preheat, e..g. "Voorverwarming van de tuit vor PLA"

File: lang/po/Firmware_nl.po [hunk 6]
```diff
@@ -0,0 +1,1787 @@
...
+
+# 
+#: ultralcd.cpp:2301
+msgid "Press the knob"
+msgstr "Druk op knop"
+
+# MSG_PRINT_PAUSED c=20 r=1
+#: ultralcd.cpp:1074
+msgid "Print paused"
+msgstr "Print pauzeren"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po:1116]: pauzeren is active, as in "pause print". Past tense ("paused") is "gepauzeerd"

File: lang/po/Firmware_nl.po [hunk 7]
```diff
@@ -0,0 +1,1787 @@
...
+
+# 
+#: ultralcd.cpp:3012
+msgid "Right"
+msgstr "Rechts"
+
+# MSG_FIND_BED_OFFSET_AND_SKEW_LINE1 c=60
+#: messages.c:38
+msgid "Searching bed calibration point"
+msgstr "Zoeken bed kalibratiepunt"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po:1236]: would suggest "Bed kalibratiepunt aan het zoeken" if there is sufficient space.

File: lang/po/Firmware_nl.po [hunk 8]
```diff
@@ -0,0 +1,1787 @@
...
+
+# 
+#: ultralcd.cpp:2308
+msgid "to load filament"
+msgstr "om filament te laden"
+
+# 
+#: ultralcd.cpp:2312
+msgid "to unload filament"
+msgstr "om filament te laden"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po:1501]: typo, should be "om filament uit te werpen"

File: lang/po/Firmware_nl.po [hunk 9]
```diff
@@ -0,0 +1,1787 @@
...
+
+# MSG_WIZARD c=17 r=1
+#: ultralcd.cpp:5814
+msgid "Wizard"
+msgstr "Wizard"
+
+# MSG_XYZ_DETAILS c=19 r=1
+#: ultralcd.cpp:2227
+msgid "XYZ cal. details"
+msgstr "XZY kal. details"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po]: XZY?

File: lang/po/Firmware_nl.po [hunk 10]
```diff
@@ -0,0 +1,1787 @@
...
+
+# 
+#: ultralcd.cpp:4031
+msgid "IR:"
+msgstr ""
+
+# 
+#: ultralcd.cpp:7152
+msgid "Magnets comp.[N/A]"
+msgstr "Magnet. comp.[N/V]"
```
Comment (vintagepc) [inline lang/po/Firmware_nl.po]: could use "Magneet" directly or "Magneten" for plural

**Review by vintagepc — COMMENTED** [review]
Reviewed and made some translation suggestions.

File: lang/po/Firmware_nl.po (hunk 8 above)
Comment (3d-gussner) [reply lang/po/Firmware_nl.po:1501]: Have to double check but i think this message is limited to c=20 r=1
Your translation is correct but is too long.

File: lang/po/Firmware_nl.po (hunk 4 above)
Comment (3d-gussner) [reply lang/po/Firmware_nl.po]: Is te lang met c=22 r=1

File: lang/po/Firmware_nl.po (hunk 9 above)
Comment (3d-gussner) [reply lang/po/Firmware_nl.po]: Typical typo... glad you reviewed it

**Comment by 3d-gussner** [comment]
@vintagepc Thanks for the comments, I will review them.
It would be better to comment on the `lang_en_nl.txt` file as these files are used during firmware build. The `.po` files are ment to exchange translations between `code` and translators.
Another thing we have to keep in mind is that the LCD can only display 20 char in 4 rows.
One of my next tasks will be to add the missing c=xx r=yy comments to be able to see what limits we have.

This means:
- one line messages may not be longer than 20 chars
- menu messages are even more limited in length
- some status messages are really short
- it is sometimes hard to use correct plural
- some words are just too long (German has some really long words)
- having a multiple row message, we still need to see how it will be shown.
  - A long message will be split into rows while it tries to get as many as possible words per row, this can look very odd if you use very long words after each other.

**Comment by vintagepc** [comment]
Understood, I knew we had space limitations but wasn't aware of the exact allowances for each case. I'm also quite aware that many EU languages can be very regional in choices of phrasing and syntax, further complicating the mess. :)

File: lang/po/Firmware_nl.po (hunk 8 above)
Comment (vintagepc) [reply lang/po/Firmware_nl.po:1501]: "om fil. te ontladen" should fit, probably better

File: lang/po/Firmware_nl.po (hunk 4 above)
Comment (vintagepc) [reply lang/po/Firmware_nl.po]: "Fil. word ontladen"?

**Comment by leptun** [comment]
There are also languages where if you keep shortening words, the message stops making sense. This is the case for Romanian. Things are also complicated by lack of proper translation counterpart. Most words either sound forced (the english counterpart is a better word that most people will understand and is the actual terminology used by the community) or are too long (ie 20+ chars for a full word).

**Comment by vintagepc** [comment]
Arabic is a fun one to deal with in translation... there if you shorten a word, it can actually take up more characters than before!

**Comment by leptun** [comment]
Luckily we don’t have to deal with these languages. The displays can’t display something like that unless it’s higher resolution and the cells are next to eachother (or there aren’t cells at all). Also, that is only an edge case, words usually get smaller in arabic just like in the other languages.

File: lang/po/Firmware_nl.po (hunk 7 above)
Comment (jurriaan) [reply lang/po/Firmware_nl.po:1236]: Bed isn't an adjective here, so it can be Bedkalibratiepunt I guess

File: Firmware/language.c [hunk 11]
```diff
@@ -209,6 +209,7 @@ const char* lang_get_name_by_code(uint16_t code)
 	case LANG_CODE_ES: return _n("Espanol");
 	case LANG_CODE_FR: return _n("Francais");
 	case LANG_CODE_IT: return _n("Italiano");
+	case LANG_CODE_NL: return _n("Nederlands");
```
Comment (DRracer) [inline Firmware/language.c]: Please tag Dutch as community contribution ...

File: lang/po/Firmware_nl.po (hunk 2 above)
Comment (3d-gussner) [reply lang/po/Firmware_nl.po]: "Auto. laden van fil. is enkel beschikbaar wanneer fil.sensor is ingeschakeld ..."
This fits perfectly the 20x4

File: lang/po/Firmware_nl.po (hunk 10 above)
Comment (3d-gussner) [reply lang/po/Firmware_nl.po]: Limited space

**Comment by 3d-gussner** [comment]
@vintagepc @jurriaan Thanks for the feedback, I updated few translations.
Can you please review the `lang_en_nl.txt` and comment if something is wrong?

**Comment by 3d-gussner** [comment]
@vintagepc Can you please review the changes https://github.com/prusa3d/Prusa-Firmware/pull/2223/commits/afa84b15bdaf09b837d091e60ab6613e943fb21d ?

**Comment by 3d-gussner** [comment]
> Maybe "De printer drukt nu een zigzaglijn. Draai..." ? Feels like it reads a little more naturally.

We have to choose between "print" and "drukt" to keep wording similar. There are over places where we use "print" instead of "drukt"

**Comment by 3d-gussner** [comment]
> Not sure if this is a regional dialect thing but I think "Riem" is more of a clothing-type belt; I've always heard "snaar" used for "technical" type belts like those in tape decks etc.

https://www.lexilogos.com/english/dutch_translation.htm 3x "riem" once "gordel"

**Comment by DRracer** [comment]
@3d-gussner @vintagepc another almost finished feature ready to be tested. 
I'd really like to see this this up and running since we managed to overcome the 64K limit in the XFLASH and verified, that it works

**Review by DRracer — APPROVED** [review]

**Pull request merged**